Entities: Adding the build entity

enoodle opened the pull request:
The build entity in ManageIQ representing the BuildConfig from Openshift.

Signed-off-by: Erez Freiberger <efreiber@redhat.com>

enoodle commented:
@simon3z please review.

simon3z commented:
@enoodle it seems that some information are missing:

https://github.com/openshift/origin/blob/v1.1/pkg/build/api/v1/types.go#L24-L50

`ServiceAccount`, `Source`, `Output`, `Resources`, `CompletionDeadlineSeconds` seem interesting enough to collect.

For the rest, @moolitayer @zeari please review.

alongoldboim commented on app/models/ems_refresh/save_inventory_container.rb:
```diff
@@ -398,4 +399,23 @@ def save_node_selector_parts_inventory(entity, hashes, target = nil)
+    ems.container_builds(true)
+    deletes = if target.kind_of?(ExtManagementSystem)
+                ems.container_builds.dup
+              else
```
This is a good spot for ternary expression:
deletes = target.kind_of?(ExtManagementSystem) ? ems.container_builds.dup : []
makes it a lot clearer and more readable.

zeari commented:
Overall OK but missing a few things:

> ServiceAccount, Source, Output, Resources, CompletionDeadlineSeconds seem interesting enough to collect.

1) In this patch I would collect anything that doesn't require another new entity.

2) This needs tests. They probably should be in ```spec/models/manageiq/providers/kubernetes/container_manager/refresh_parser_spec.rb```
since making a new VCR for each in ```spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb``` is too much.

3) The Travis failure looks legit.

enoodle commented:
@zeari As we discussed, The Travis problem will be fixed after a new version of openshift_client is available and we will know the new version number to update the GemFile.
I added ServiceAccount, Source, Output, CompletionDeadlineSeconds but still have Resources to parse but I had some trouble with their definition. so far I added all but the Output values to the DB.
@moolitayer would you like to review too?

zeari commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -47,6 +52,71 @@ def parse_route(route)
+
+      def parse_source_secret(secret)
+        secret ? secret.name : nil
+      end
```
why not use ```try``` here?
http://apidock.com/rails/Object/try

zeari commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -47,6 +52,71 @@ def parse_route(route)
+        secret ? secret.name : nil
+      end
+
+      def parse_build_source(source_item)
```
Could ```source_item``` be ```nil```?

zeari commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -47,6 +52,71 @@ def parse_route(route)
+          :build_source_type => source_item.type,
+          :binary            => parse_binary_source(source_item.binary),
+          :dockerfile        => source_item.dockerfile,
+          :git               => parse_git_source(source_item.git).try(:uri),
```
If only ```uri``` interests you here then I would do something like:
```Ruby
source_item.try(:git).try(:uri),
```

I think that having something like ```.try(:source_item).try(:git).try(:uri)``` isn't so bad.
what do you think @moolitayer ?

moolitayer replied:
@zeari not sure what you mean, what about parse_git_source

moolitayer commented:
@enoodle as discussed there should be an increment of the openshift_client gem in this patch[1], 
you can mark this as wip until such version is made (using ```@miq-bot add_label wip```)

[1] https://github.com/ManageIQ/manageiq/blob/master/gems/pending/Gemfile

moolitayer commented:
to add the new entities also update the list at:
https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/openshift/container_manager/refresher.rb

moolitayer commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -47,6 +52,71 @@ def parse_route(route)
+          result.merge!(key => ar_item.try(key))
+        end
+        result
+      end
```
```
def parse_strings(ar_item, keys)
  keys.each_with_object({}) do |key, h|
    h[key] = ar_item.try(key)
  end
end
```

moolitayer commented on app/models/container_build.rb:
```diff
@@ -0,0 +1,14 @@
+class ContainerBuild < ActiveRecord::Base
+  include CustomAttributeMixin
+  include ReportableMixin
```
I'm not sure we need CustomAttributeMixin, ReportableMixin, has_many :lables and acts_as_miq_taggable.
we can take a look on Sunday. maybe @zeari knows

zeari replied:
> I'm not sure we need CustomAttributeMixin, ReportableMixin, has_many :lables and acts_as_miq_taggable.
> we can take a look on Sunday. maybe @zeari knows?

@moolitayer 
```CustomAttributeMixin``` for labels\selectors - OK
```ReportableMixin``` for producing reports - should be added with a PR that adds a reports feature for builds
```acts_as_miq_taggable``` for tags - should be added with a PR that lets the user add tags to a build.

If any of the last two are coming in subsequent patch then leave them in.

moolitayer replied:
@enoodle please remove  
```include ReportableMixin```

enoodle replied:
@moolitayer I tried but apparently we need all three of those for some of the screens in the UI.

moolitayer replied:
ok thanks

simon3z commented on db/migrate/20151216095054_create_container_builds.rb:
```diff
@@ -0,0 +1,22 @@
+      t.string     :build_source_type
+      t.string     :binary
+      t.string     :dockerfile
+      t.string     :git
```
@enoodle I prefer the naming to be scoped, for example instead of `git` it should probably be `source_git`, same for the others (`context_dir`, etc.).

simon3z commented on db/migrate/20151216095054_create_container_builds.rb:
```diff
@@ -0,0 +1,22 @@
+      t.string     :namespace
+      t.string     :service_account
+      t.string     :build_source_type
+      t.string     :binary
```
@enoodle reading in the doc this seems to be a tar file (I could be wrong, please double check). If that's the case then it would be overkill to persist it (don't collect it).

enoodle replied:
@simon3z I understand that its the file name of a binary file. Its not 100% clear but the do write that it "must be a valid filename with no path segments". do you think its not important to save right now?

enoodle commented:
@miq-bot add_label wip

simon3z commented:
Overall direction is good (I didn't go into details), please @moolitayer @zeari review. Thanks!

simon3z commented:
@enoodle can you split out the UI side of this PR? Thanks.

enoodle commented:
@simon3z Split, The UI is in #6084

simon3z commented:
At first sight this looks in a good state. @moolitayer @zeari @dkorn @alongoldboim can you review?

zeari commented:
Aside from styling errors I think this is ok.

zeari commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -47,6 +65,82 @@ def parse_route(route)
+          :completion_deadline_seconds => build.spec.try(:completionDeadlineSeconds),
+        )
+
+        new_result[:output_name] = parse_build_output(build.spec.try(:output))[:to][:name]
```
@enoodle ```parse_build_output(build.spec.try(:output))``` could be null

enoodle replied:
Right, but ```parse_build_output``` will return a valid hash with ```:to`` that has a ```:name``` anyway.

zeari commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -47,6 +65,82 @@ def parse_route(route)
+          :output_docker_image_reference => status[:OutputDockerImageReference],
+        )
+        new_result[:build_config] = @data_index.fetch_path(:container_builds, :by_name,
+                                                           build_pod.status.config.try(:name))
```
@enoodle can ```build_pod.status.config.try(:name)``` return nil?

enoodle replied:
Yes, but it will be OK: fetch_path will not find a key nil, will return nil and this is the expected result.

enoodle commented:
@miq-bot remove_label wip

moolitayer commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -47,6 +65,82 @@ def parse_route(route)
+
+      def parse_strings(ar_item, keys)
+        keys.each_with_object({}) do |key, h|
+          h[key] = ar_item.try(key)
```
is calling try any different then

```h[key] = ar_item[key]```

moolitayer commented on app/models/ems_refresh/save_inventory_container.rb:
```diff
@@ -398,4 +402,35 @@ def save_node_selector_parts_inventory(entity, hashes, target = nil)
+      h[:container_project_id] = h.fetch_path(:project, :id)
+    end
+    save_inventory_multi(ems.container_builds, hashes, deletes, [:ems_ref], [],
+                         [:labels, :project, :resources])
```
If you indeed want to save labels for container_builds as you defined in your model classes (has_many :labels)
you should probably pass them as child_keys and not ignore_keys

enoodle commented:
@miq-bot add_label wip

simon3z commented:
> @miq-bot add_label wip

@enoodle why is this a WIP? What is missing?

enoodle commented:
@simon3z I want to add a relation to container_image

simon3z commented:
> @simon3z I want to add a relation to container_image

if you address that in another PR, would this be ready now?

enoodle commented:
@simon3z yes

simon3z commented:
Then this LGTM :+1: cc @chessbyte ready for merge.

@enoodle please remove the wip label and address the `container_image` relationship in another PR.

enoodle commented:
@miq-bot remove_label wip

blomquisg commented:
@simon3z good call separating out UI to another PR ... much easier to look at this in isolation :smile:
